lib/rule.js

```diff
@@ -236,43 +236,84 @@ Rule.prototype = new (function () {
       , source
       , action
       , opts
-      , prereqs
       , parts
       , valid
       , src
       , tNs
       , createdTask
       , name = Task.getBaseTaskName(fullName)
       , nsPath = Task.getBaseNamespacePath(fullName)
-      , ns = this.ns.resolveNamespace(nsPath);
+      , ns = this.ns.resolveNamespace(nsPath)
+      , prereqs = this.prereqs.slice(); // Get a copy to work with
 
     pattern = this.pattern;
     source = this.source;
 
-    if (typeof source == 'string') {
-      src = Matcher.getSource(name, pattern, source);
+    // TODO: Write a utility function that appends a
+    // taskname to a namespace path
+    var completeNamespace = function(s){
+      return nsPath.split(':').filter(function (item) {
+        return !!item;
+      }).concat(s).join(':');
+    };
+
+    // Deal with the various things the source argument can be:
+    // 1: A function
+    if (typeof source == 'function') { 
+      source = source(name);
+
+      if (source instanceof Array) {
```

Comment:
Looks like duplcate logic of what's below. Can you refactor this so that it does its function stuff first, then do the regular logic for Array/String?

Reply:
I had a similar thought - but then noticed each instance works slightly differently.

The `if string` outside of `if function` uses the matcher:

```
src = completeNamespace( Matcher.getSource(name, pattern, source) )
```

while `if string` inside of `if function` just uses `completeNamespace`:

```
src = completeNamespace(source);
```
(same for `if array`)

From skimming the source, I think that means if a string is passed directly, it can contain wildcards, but if it's returned from a function, it must be the entire filename of the source, without wildcards, etc.

Now, maybe wildcards should always be applied (or I've misinterpreted the code I skimmed), but that's my 2 cents.